gtime/hierarchical/bottom_up.py

```diff
@@ -0,0 +1,145 @@
+            A dictionary of time series. Each is fitted independently
+        y : pd.DataFrame, optional, default = ``None``
+            only for compatibility
+
+        Returns
+        -------
+        self
+        """
+        self._check_is_dict_of_dataframes_with_str_key(X)
+        if self.hierarchy_tree == 'infer':
+            self._infer_hierarchy_tree(X)
+        super()._initialize_models(X)
+        for key, time_series in X.items():
+            self.models_[key].fit(time_series)
+        return self
+
+    def predict(self, X: Dict[str, pd.DataFrame] = None):
+        """ Predict method
+
+        Parameters
+        ----------
+        X : Dict[str, pd.DataFrame], optional, default = ``None``
+            time series to predict. If ``None`` all the fitted time series are predicted.
+            The keys in ``X`` have to match the ones used to fit.
+
+        Returns
+        -------
+        predictions : Dict[str, pd.DataFrame]
+        """
+        check_is_fitted(self)
+        if X is None:
+            return self._predict_fitted_time_series()
+        else:
+            return super()._predict_new_time_series(X)
+
+    def _infer_hierarchy_tree(self, X: Dict[str, pd.DataFrame]):
+        self.hierarchy_tree = nx.DiGraph()
+        for key in list(X.keys()):
+            self.hierarchy_tree.add_node(key)
+        for key in list(X.keys())[1:]:
+            self.hierarchy_tree.add_edge(list(X.keys())[0], key)
+        # First key of X is the root, all other keys are leaves of the tree
+
+    def _predict_fitted_time_series(self) -> Dict[str, pd.DataFrame]:
+        bottom_up_dictionary = {}
+        for key, model in self.models_.items():
+            self._bottom_up_addiction_to_dictionary(key, model, bottom_up_dictionary)
+        return bottom_up_dictionary
+
+    def _bottom_up_addiction_to_dictionary(self, key, model, dict_bottom_up):
+        if self._is_a_leaf(key):
+            dict_bottom_up[key] = model.predict()
+        else:
+            self._check_predict_children_computed(dict_bottom_up, key)
+            self._sum_children_prediction(key, dict_bottom_up)
+
+    def _is_a_leaf(self, key) -> bool:
+        if self.hierarchy_tree.out_degree[key] == 0:
+            return True
+        else:
+            return False
+
+    def _sum_children_prediction(self, parent_key, dict_bottom_up):
+        temp = deepcopy(dict_bottom_up[list(self.hierarchy_tree[parent_key])[0]])
+        for item in list(self.hierarchy_tree[parent_key])[1:]:
+            temp = temp + dict_bottom_up[item]
+        dict_bottom_up[parent_key] = temp
+        return
+
+    def _check_predict_children_computed(self, dict_bottom_up, tested_key):
+        for child in list(self.hierarchy_tree[tested_key]):
+            if child not in dict_bottom_up:  # If child model not computed yet
+                if self._is_a_leaf(child):
+                    dict_bottom_up[child] = self.models_[child].predict()
```

Comment:
Once a prediction is done one time or it is recovered by the algorithm it will appear in the dict_bottom_up so the computation will be done just once.